Update the cycle point icon state

kinow opened the pull request:
These changes close #666

@oliver-sanders could you test and see if this solves the issue for your workflow? I tested on `five`, and noticed that when the job started running, the cycle point icon never changed. Then, after this change, it was correctly updated. Thanks!

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Appropriate tests are included (unit and/or functional).
- [x] Appropriate change log entry included.
- [x] No documentation update required.

kinow commented:
Looking how hard it would be to write a test for this :+1: it should be covered by existing tests, but not necessarily verifying whether the icon is updated when a state changes after the page & component have been loaded. Right now I'm thinking about simply `setTimeout` to change a state in some JS object after ~5 secs, and then mocha or Cypress asserting it changed in the DOM element as well (`nextTick`).

hjoliver commented on src/components/cylc/tree/cylc-tree.js:
```diff
@@ -67,7 +67,8 @@ function computeCyclePointsStates (cyclePointNodes) {
     }
-    cyclePointNode.node.state = extractGroupState(childStates, false)
+    const cyclePointState = extractGroupState(childStates, false)
+    Vue.set(cyclePointNode.node, 'state', cyclePointState)
```
(I had been wondering about `Vue.set`; nice doc here: https://vuejs.org/v2/api/#Vue-set )

oliver-sanders commented:
Yes this fixes it :+1:.

kinow commented:
Thanks for testing it! I will try to come up with a test to prevent it from happening again (in Vue 3 you just create an object like `const cyclePoints = ref(someObject)` or `... = reactive(someObject)` and in theory reactivity just-works™).

oliver-sanders commented:
Why was this particular property non-reactive?

kinow commented:
> Why was this particular property non-reactive?

Ah, good question! The answer is in the docs @hjoliver linked

>Adds a property to a reactive object, ensuring the new property is also reactive, so triggers view updates. This must be used to add new properties to reactive objects, as Vue cannot detect normal property additions (e.g. this.myObject.newProperty = 'hi').

We have a `cyclePointNode.node` object. The `.node` part, if I recall correctly, contains the GraphQL object as it was returned from the GraphQL endpoint. That object is reactive I think, but the Cycle Point (a [family proxy](https://github.com/cylc/cylc-ui/blob/2042834d65f00ca9a77db1836066c56397221d1d/src/graphql/queries.js#L56-L58) actually in our query) does not contain the `.state` property.

The old code was simply adding `.state = 'bla'` in the reactive object.

But as the Vue docs text above states, that does not mean that that attribute is reactive; so when used in templates, only the initial value is displayed. Updates to the value do not trigger an update to the template.

kinow commented:
Added a comment to the code quickly explaining why we have `Vue.set`. Added a unit test which verifies that the JS object is an observable (by confirming it has a JS [getter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/get) which is defined by Vue), and added a changelog entry.

I started writing an e2e test… but wasn't sure if that was the best way to test such a small part of the code. Reading the Vue docs and some of its code, I realized it would be simpler to test that the `node.state` was reactive, and not a simple JS property.

This means we can have a small and simple unit test, instead of an e2e test (e2e tests are harder to maintain and debug, but more useful when you want to test DOM, a real page, site navigation, etc).

Ready for review!

hjoliver reviewed (approved):
:+1:
